File: iosApp/iosApp/Views/Login/LoginViewModel.swift
//
//  LoginViewModel.swift
//  iosApp
//
//  Created by Jan Cortiel on 06.02.23.
//  Copyright © 2023 Ludwig Boltzmann Institute for
//  Digital Health and Prevention - A research institute
//  of the Ludwig Boltzmann Gesellschaft,
//  Oesterreichische Vereinigung zur Foerderung
//  der wissenschaftlichen Forschung 
//  Licensed under the Apache 2.0 license with Commons Clause 
//  (see https://www.apache.org/licenses/LICENSE-2.0 and
//  https://commonsclause.com/).
//


import shared
import KMPNativeCoroutinesCombine
import Combine

class LoginViewModel: ObservableObject {
    private let registrationService: RegistrationService
    @Published var endpoint: String = ""
    @Published var defaultEndpoint: String = ""
    @Published var token: String = ""
    
    @Published var showQRCodeView: Bool = false
    
    private var cancellables: Set<AnyCancellable> = []
    
    
    init(registration: RegistrationService) {
        self.registrationService = registration
        defaultEndpoint = registration.getEndpointRepository().endpoint()
        
        Publishers.CombineLatest($endpoint, $token)
            .map { !$0.isEmpty || !$1.isEmpty }
            .removeDuplicates()
            .sink { [weak self] hasInput in
                self?.registrationService.clearError()
            }
            .store(in: &cancellables)
    }
    
    func extractValuesFromQRCode(qrCodeUrl: String) {
        endpoint = qrCodeUrl.components(separatedBy: "signup?").first ?? ""
         
         if let tokenPart = qrCodeUrl.components(separatedBy: "token=").last,
               tokenPart != qrCodeUrl {
                token = tokenPart.components(separatedBy: "&").first ?? ""
         }
     }
    
    func validate() {
        let loginModel = LoginModel(token: token, endpoint: currentStudyEndpoint())
        if loginModel.valid() {
            registrationService.sendRegistrationToken(loginModel: loginModel)
        } else {
            AlertController.shared.openAlertDialog(model: AlertDialogModel(title: "Token or Endpoint invalid", message: "login_model_invalid_body", positiveTitle: "Ok", negativeTitle: nil, onPositive: {
                AlertController.shared.closeAlertDialog()
            }, onNegative: nil))
        }
    }
    
    func checkTokenCount() -> Bool {
        return self.token.count == 0
    }
    
    func currentStudyEndpoint() -> String {
        endpoint.isEmpty ? defaultEndpoint : endpoint
    }
}

File: androidApp/src/main/java/io/redlink/more/app/android/activities/login/LoginViewModel.kt
/*
 * Copyright LBI-DHP and/or licensed to LBI-DHP under one or more
 * contributor license agreements (LBI-DHP: Ludwig Boltzmann Institute
 * for Digital Health and Prevention -- A research institute of the
 * Ludwig Boltzmann Gesellschaft, Österreichische Vereinigung zur
 * Förderung der wissenschaftlichen Forschung).
 * Licensed under the Apache 2.0 license with Commons Clause
 * (see https://www.apache.org/licenses/LICENSE-2.0 and
 * https://commonsclause.com/).
 */
package io.redlink.more.app.android.activities.login

import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.ViewModel
import io.redlink.io.more.app.android.R
import io.redlink.more.AlertController
import io.redlink.more.app.android.extensions.stringResource
import io.redlink.more.models.AlertDialogModel
import io.redlink.more.models.LoginModel
import io.redlink.more.registration.RegistrationService
import io.redlink.more.util.validateAndNormalizeUrl

class LoginViewModel(
    val registrationService: RegistrationService
) : ViewModel() {
    val participantKey = mutableStateOf("")
    val isLoading = registrationService.isLoading
    val error = registrationService.error

    val dataEndpoint = mutableStateOf("")
    val defaultEndpoint = mutableStateOf(registrationService.getEndpointRepository().endpoint())
    val endpointError = mutableStateOf<String?>(null)

    fun participationKeyNotBlank(): Boolean = this.participantKey.value.isNotBlank()
    fun isEndpointError(): Boolean =
        !this.endpointError.value.validateAndNormalizeUrl().isNullOrBlank()

    fun currentEndpoint() = dataEndpoint.value.ifEmpty { defaultEndpoint.value }

    fun validateKey() {
        if (registrationService.connected.value) {
            val loginModel = LoginModel(participantKey.value, currentEndpoint())
            if (loginModel.valid()) {
                registrationService.sendRegistrationToken(loginModel)
            } else {
                AlertController.openAlertDialog(
                    AlertDialogModel(
                        stringResource(R.string.more_token_error),
                        stringResource(R.string.more_404),
                        confirmLabel = "Ok"
                    )
                )
            }
        } else {
            val dialogModel = AlertDialogModel(
                title = stringResource(R.string.no_internet_connection_title),
                message = stringResource(R.string.no_internet_connection_body),
                confirmLabel = "Ok"
            )
            AlertController.openAlertDialog(dialogModel)
        }
    }

    fun extractValuesFromQRCode(qrCodeUrl: String) {
        dataEndpoint.value =
            qrCodeUrl.substringBefore("signup?").validateAndNormalizeUrl() ?: currentEndpoint()
        participantKey.value =
            qrCodeUrl.substringAfter("token=", "").takeIf { it.isNotEmpty() } ?: ""
    }
}